pkg/backend/filestate/gcpauth.go

```diff
@@ -0,0 +1,77 @@
+package filestate
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/util/cmdutil"
+
+	"golang.org/x/oauth2/google"
+
+	"gocloud.dev/blob/gcsblob"
+
+	"cloud.google.com/go/storage"
+
+	"github.com/pkg/errors"
+	"gocloud.dev/blob"
+	"gocloud.dev/gcp"
+)
+
+type GoogleCredentials struct {
+	PrivateKeyID string `json:"private_key_id"`
+	PrivateKey   string `json:"private_key"`
+	ClientEmail  string `json:"client_email"`
+	ClientID     string `json:"client_id"`
+}
+
+func GoogleCredentialsMux() (*blob.URLMux, error) {
+
+	var err error
+	var credentials *google.Credentials
+
+	if creds := os.Getenv("GOOGLE_CREDENTIALS"); creds != "" {
+		// We try $GOOGLE_CREDENTIALS before gcp.DefaultCredentials
+		// so that users can override the default creds
+		credentials, err = google.CredentialsFromJSON(context.TODO(), []byte(creds), storage.ScopeReadWrite)
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to parse credentials from $GOOGLE_CREDENTIALS")
+		}
+	} else {
+		// DefaultCredentials will attempt to load creds in the following order:
+		// 1. a file located at $GOOGLE_APPLICATION_CREDENTIALS
+		// 2. application_default_credentials.json file in ~/.config/gcloud or $APPDATA\gcloud
+		credentials, err = gcp.DefaultCredentials(context.TODO())
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to find gcp credentials")
+		}
+	}
+
+	client, err := gcp.NewHTTPClient(gcp.DefaultTransport(), credentials.TokenSource)
+	if err != nil {
+		return nil, err
+	}
+
+	options := gcsblob.Options{}
+
+	account := GoogleCredentials{}
+	err = json.Unmarshal(credentials.JSON, &account)
+	if err == nil && account.ClientEmail != "" && account.PrivateKey != "" {
+		options.GoogleAccessID = account.ClientEmail
+		options.PrivateKey = []byte(account.PrivateKey)
+	} else {
+		cmdutil.Diag().Warningf(diag.Message("",
+			"Pulumi will not be able to pint a statefile permalink using these credentials. "+
+				"Neither a GoogleAccessID or PrivateKey are available. "+
+				"Try using a GCP Service Account."))
```

Comment:
these feel bad that thse are warnings.  Why don't we just require this stuff?  Does it make sense to allow people to use these guys if we can't even get the credentials for them?

Reply:
Well, if you're using pulumi in CI then yeah it makes sense that the most common use case is a service account which has the right information in the google credential file for creating signed urls.

But for human usage people are likely authenticating via `gcloud auth login` which will setup a credentials file for them in `~/.config/gcloud/` of type `"type": "authorized_user"` which doesn't have the required info for `SignedURL()`. I'm not sure why it all pans out like this but it does :/

In terms of actually using pulumi to provision infra to GCP as well as accessing a gs:// bucket, both authentication mechanisms work fine.

If we require this info then users will be forced to create a service account to run pulumi which is unneeded friction. Warning is preferred because it's not a critical problem - you won't see a statefile permalink, no big deal. All the infra will be created/updated and you'll see a warning telling your how to solve the issue if you really needed the permalink. I doubt anyone does anything with the permalink other than use it to quickly get to the statefile if debugging. Worse case they navigate there manually which is possible because pulumi has a simple file layout.